hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/node/NodeStateManager.java

```diff
@@ -278,10 +292,21 @@ private void initializeStateMachines() {
    * @throws NodeAlreadyExistsException if the node is already present
    */
   public void addNode(DatanodeDetails datanodeDetails,
-                      LayoutVersionProto layoutInfo)
-      throws NodeAlreadyExistsException {
+      LayoutVersionProto layoutInfo) throws NodeAlreadyExistsException {
     NodeStatus newNodeStatus = newNodeStatus(datanodeDetails);
     nodeStateMap.addNode(datanodeDetails, newNodeStatus, layoutInfo);
+    UUID dnID = datanodeDetails.getUuid();
+    try {
+      updateLastKnownLayoutVersion(datanodeDetails, layoutInfo);
```

Comment:
Do we need to add a state == HEALTHY_READONLY check here?

Reply:
Just an alternative suggestion, can we init state as HEALTHY for the Datanode, and move it to READONLY based on the alternative condition (layoutMismatch)? It is more likely for a new DN to be in a correct rather than a wrong layout version.

Reply:
Starting nodes as healthy and moving to healthy readonly when node added if there is a layout mismatch makes sense. I updated the code.

Reply:
`updateLastKnownLayoutVersion`  and `updateNodeLayoutVersionState` I believe will handle the state transitions without an extra check.